Workflows Base Module

cadeduckworth opened the pull request:
close #237

cadeduckworth commented:
This PR will require that PR#217 or PR#227 is merged to complete testing of directory iteration function and workflows registry module.

cadeduckworth commented:
Requires merging of recent updates from `develop` before continuing work on tests.

orbeckst commented:
There are test failures related to SM25/SM26. Let me know when the PR is ready for review (or if you need input on something specific).

cadeduckworth commented:
@orbeckst I added the workflows testing data path to the STATES dictionary in __init__.py and cleaned up the testing module by making the necessary changes and adding a few more fixtures. Once it is confirmed that everything looks good I will go through the docs once more to revise/finalize!

cadeduckworth commented:
UPDATE: I forgot to change the single quotes to double quotes to remove the backslash from logger/error messages in `base.py` and `test_worflows_base.py`. 

Also, the `dir_paths.csv` file in the testing resources still might need a change of paths to a relative convention, since it assumes that the CI is running the tests from outside of the `mdpow` directory using an absolute path, and paths can always change, which would break the test.

orbeckst reviewed (changes requested):
Looking really good, I don't see any problems.

EDIT: Please also add some documentation for the registry.

Just add an Enhancement entry to CHANGES. (Please also fix the year of the 0.8.0 release to 2022 instead of 2021).

cadeduckworth commented:
@orbeckst 

requested changes addressed

orbeckst commented on mdpow/workflows/registry.py:
```diff
@@ -0,0 +1,35 @@
+:class:`~mdpow.analysis.ensemble.EnsembleAnalysis` should be iterated over
+the provided project data.
+
+.. seealso:: :mod:`~mdpow.workflows.base`
```
make this a single seealso (it can do lists, check syntax)

orbeckst commented on mdpow/workflows/registry.py:
```diff
@@ -0,0 +1,35 @@
+=======================================================================================
+
+Each entry in :mod:`mdpow.workflows.registry` corresponds to an :class:`~mdpow.analysis.ensemble.EnsembleAnalysis` for which exists a corresponding
+automated workflow.
```
Here should be a list of workflows, essentially manually maintain what's in registry.

orbeckst commented on mdpow/workflows/registry.py:
```diff
@@ -0,0 +1,35 @@
+ 
+registry = {
+    'DihedralAnalysis' : dihedrals.automated_dihedral_analysis,
+    'SolvationAnalysis' : None,
+    'HBondAnalysis' : None
```
remove anything that's not present yet

orbeckst commented on mdpow/workflows/registry.py:
```diff
@@ -0,0 +1,35 @@
+# import analysis
+from mdpow.workflows import dihedrals
+ 
+registry = {
```
State here for developers what needs to be added, i.e., function/class

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,166 @@
+       .. rubric:: Example
+       
+       Typical Workflow::
+
+           directory_paths = directory_paths(parent_directory='/foo/bar/MDPOW_projects')
+           directory_iteration(directory_paths)
+
+           or
+
+           directory_paths = directory_paths(csv='/foo/bar/MDPOW.csv')
+           directory_iteration(directory_paths)
```
do proper code formatting

cadeduckworth replied:
Also, can you look at the `.. autodata:: registry` portion of registry.py and let me know if there is a way to suppress the type hint. If you look at the docs build it included instructions on how to build a dictionary underneath the dictionary values for registry.

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,166 @@
+            logger.info(f'dir_paths saved under {csv_save_dir}')
+
+        else:
+            current_directory = os.system('pwd')
```
Don't use `os.system` — use a python function such as `cwd` or something from `pathlib`

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,166 @@
+    """Takes a :class:`pandas.DataFrame` created by
+       :func:`~mdpow.workflows.base.directory_paths`
+       and iterates over the project paths to implement
+       :func:`~mdpow.workflows.dihedrals.automated_dihedral_analysis`
```
write docs more generally, not just dihedrals

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,166 @@
+           that corresponds to the desired automated workflow module
+
+       *kwargs*
+           keyword arguments from :func:`~mdpow.workflows.dihedrals.automated_dihedral_analysis`
```
write docs more generally, not just dihedrals

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,166 @@
+def directory_iteration(directory_paths, ensemble_analysis, **kwargs):
+    """Takes a :class:`pandas.DataFrame` created by
+       :func:`~mdpow.workflows.base.directory_paths`
+       and iterates over the project paths to implement
```
not "implement" but "run"

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,166 @@
+"""
+:mod:`mdpow.workflows.base` --- Automated workflow base functions
+=================================================================
+
```
Add text to explain (primarily for developers) what you need to do to make this module work. Explain (and link) to registry.

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,166 @@
+
+    return directory_paths
+
+def directory_iteration(directory_paths, ensemble_analysis, **kwargs):
```
The name `directory_iteration` is not telling me that we are running an EnsembleAnalysis for multiple MDPOW runs. Come up with a better name.

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,166 @@
+
+        raise TypeError(msg)
+
+    return logger.info('all analyses completed')
```
Returning logger does not make sense.

I think we should be returning a collection of the data that's produced — eg all filenames. How will you know after this has run, what you got? It should _not_ require knowing a file naming scheme.

orbeckst replied:
For right now, please change to 
```python
logger.info('all analyses completed')
return
```
(and elsewhere, too)

orbeckst reviewed (changes requested):
I found a bunch of doc things and small code stuff.

EDIT (after our discussion): Please address the "small stuff" in the comments. For the bigger issue below (**maintaining a database of all produced artifacts/results**), please _open a new issue_:

The bigger conceptual issue is that with the current automated workflows we dump a whole bunch of files into the file system without recording them. I would like your automated analysis to return a manifest of files so that downstream code can work with these data structures to do other things — eg compile them into PDFs, load CSVs for further processing etc. It shouldn't be necessary to reconstruct file names by rules that are embodied inside your code.

This means that any automated workflow should be ultimately returning something like a dict or list (or df) of "type_of_output -> filename". 

The `directory_iteration()` should then bundle everything together, probably in a big dict or better a df that allows one to "get all dihedral KDE plots that contain octanol" — I imagine a df with many columns (molecule, solvent, quantity, datatype,... , filename) where datatype could be "image" or "data" (for CSV) or similar. As part of `base` you'd have to _define_ this API.

cadeduckworth commented:
@orbeckst 

I think I properly addressed all of the requested changes. I will come back and review the documents with fresh eyes later. Also, see question above.

orbeckst commented:
@cadeduckworth can you please update the branch with the main branch and please raise the issue about generating manifest dataframe for all generated data/images.

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,180 @@
+:class:`~mdpow.analysis.ensemble.EnsembleAnalysis` for all MDPOW projects
+under the top-level directory provided to :func:`project_paths`.
+
+Currently supported workflows - :mod:`~mdpow.workflows.dihedrals`
```
not needed here
```suggestion
```

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,180 @@
+            'resname': resnames,
+            'path': locations
+        }
+    )
```
indent

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,180 @@
+        }
+    )
+        if csv_save_dir is not None:
+            
```
can remove some blank lines
```suggestion
```

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,180 @@
+            
+            project_paths.to_csv(f'{csv_save_dir}/project_paths.csv', index=False)
+            logger.info(f'project_paths saved under {csv_save_dir}')
+
```
make a bit more compact
```suggestion
```

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,180 @@
+
+       .. rubric:: Example
+       
+       Typical Workflow::
+
+           project_paths = project_paths(parent_directory='/foo/bar/MDPOW_projects')
+           automated_project_analysis(project_paths)
+
+           'or'
+
+           project_paths = project_paths(csv='/foo/bar/MDPOW.csv')
+           automated_project_analysis(project_paths)
+
```
If you look at https://mdpow--229.org.readthedocs.build/en/229/workflows/base.html#mdpow.workflows.base.project_paths
![image](https://user-images.githubusercontent.com/237980/228389825-8d075ef0-abfa-4458-89e1-d496bf53e935.png)
you see that the "or" is still formatted as code. 

The formatting below should fix it:

```suggestion
       Typical Workflow::
           project_paths = project_paths(parent_directory='/foo/bar/MDPOW_projects')
           automated_project_analysis(project_paths)
           
        or::
        
           project_paths = project_paths(csv='/foo/bar/MDPOW.csv')
           automated_project_analysis(project_paths)
           
```

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,180 @@
+           that corresponds to the desired automated workflow module
+
+       *kwargs*
+           keyword arguments for the supported automated workflows,
+           currently - :func:`~mdpow.workflows.dihedrals.automated_dihedral_analysis`
+
+           .. autodata:: mdpow.workflows.dihedrals.automated_dihedral_analysis
```
remove the details on automated_dihedral_analysis
```suggestion
           keyword arguments for the supported automated workflows, see the
           :mod:`~mdpow.workflows.registry` for all available workflows and their
           call signatures
```
(also, the `autodata` messes up formatting — I am not sure what you actually wanted to show here, if you want to link, just do a 
```rest
:func:`mdpow.workflows.dihedrals.automated_dihedral_analysis`
```

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -0,0 +1,180 @@
+
+            logger.info(f'starting {molname}')
+
+            registry.registry[ensemble_analysis](dirname=dirname, resname=resname, molname=molname, **kwargs)
```
Wrap _only_ this line in try/except.

The smaller the code that you're bracketing, the more precisely you are still able to locate other errors.

orbeckst commented on mdpow/workflows/registry.py:
```diff
@@ -0,0 +1,31 @@
+Each entry in :mod:`mdpow.workflows.registry` corresponds to an :class:`~mdpow.analysis.ensemble.EnsembleAnalysis`
+for which exists a corresponding automated workflow.
+
+Currently supported workflows - :mod:`~mdpow.workflows.dihedrals`
```
Does not format as a list, see https://mdpow--229.org.readthedocs.build/en/229/workflows/registry.html#mdpow-workflows-registry-registry-of-currently-supported-automated-workflows
![image](https://user-images.githubusercontent.com/237980/228390948-05ac5cf2-a3ac-4e0e-8d80-ad32087e9bae.png)

Can you please create a [reST table](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#tables) (and see more under [table role](https://docutils.sourceforge.io/docs/ref/rst/directives.html#table)) that looks like

| **keyword** | **analysis function** |
|---------|---------|
| DihedralAnalysis | :func:'dihedrals.automated_dihedral_analysis <mdpow.workflows.dihedrals.automated_dihedral_analysis>' |

where the keyword is the string needed for :func:'mdpow.workflows.base.automated_project_analysis' and the analysis function is the actual automated function that will be executed — essentially the same information that is in the registry dict but immediately understandable (and usable) by humans.

For the reST, should look something like (include [name](https://docutils.sourceforge.io/docs/ref/rst/directives.html#common-options) for hyperlinking)
```reST
.. table:: Available automated workflows.
   :widths: auto
   :name: workflows_table

    ==============  ============
    keyword         function
    ==============  ============
    ...
    ...
    ==============  ============
```

orbeckst commented on mdpow/workflows/registry.py:
```diff
@@ -0,0 +1,31 @@
+:class:`~mdpow.analysis.ensemble.EnsembleAnalysis`, with the value defined as
+`<workflow module name>.<top-level function>`.
+
+.. autodata:: registry
```
you can probably use `:nomember:` or other options — check the docs https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html

orbeckst commented on mdpow/workflows/registry.py:
```diff
@@ -0,0 +1,31 @@
+
+.. autodata:: registry
+
+.. seealso:: :mod:`~mdpow.workflows.base`, :mod:`~mdpow.workflows.dihedrals`
```
only include base

orbeckst commented on mdpow/workflows/registry.py:
```diff
@@ -0,0 +1,31 @@
+# import analysis
+from mdpow.workflows import dihedrals
+ 
+registry = {
```
Add documentation for the `registry`. Either as part of the code
```python
#: The `registry` contains ... 
#: New workflows are added by ...
#:    The currently available workflows are listed in  the :ref:`Table of Workflows <workflows_table>`
registry = { ... }
```
or by adding the text above when writing 
```rest
.. :data:`registry`
   The `registry` contains ...   New workflows are added by ...
   The currently available workflows are listed in  the :ref:`Table of Workflows <workflows_table>`
```

(note the hyperlink ref to the table's name)

cadeduckworth replied:
@orbeckst 

I can't figure this part out. I am having trouble writing the `..  data::` or `:data:` section in a way that is recognized by `.. autodata::`.

The docs are very confusing when it comes to documenting objects within a module.

orbeckst reviewed (changes requested):
Please see comments for requested improvements. Thanks!

orbeckst commented:
@cadeduckworth please ping me when you're ready for review

cadeduckworth commented:
@orbeckst 
ready for review

orbeckst reviewed (approved):
looks good, let me just merge to doc changes.

orbeckst commented:
Hooray 🎉 , it's merged @cadeduckworth !
